File: general-interview-prep/javascript/rotateNxNMatrix.js
let oldNNMatrix =
  [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
  ];

// Works for NxN matrixes
const rotateNxNMatrix = (matrix) => matrix.map((column, element) => matrix.map((row) => row[element]).reverse());

console.log(rotateNxNMatrix(oldNNMatrix));



/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    for (let i = 0; i < matrix.length / 2; i++) {
        for (let j = i; j < matrix.length - i - 1; j++) {
            // Swap elements of each cycle
            // in clockwise direction
            let temp = matrix[i][j];
            matrix[i][j] = matrix[matrix.length - 1 - j][i];
            matrix[matrix.length - 1 - j][i] = matrix[matrix.length - 1 - i][matrix.length - 1 - j];
            matrix[matrix.length - 1 - i][matrix.length - 1 - j] = matrix[j][matrix.length - 1 - i];
            matrix[j][matrix.length - 1 - i] = temp;
        }
    }
    return matrix;
};

File: general-interview-prep/golang/rotateNXNMatrix.go
package main

import (
    "fmt"
)

// reverse diagonal elements first, then reverse each row
func rotate(matrix [][]int) {
    size := len(matrix)
	for i := 1; i < size; i++ {
		for j := 0; j < i; j++ {
			tmp := matrix[i][j]
			matrix[i][j] = matrix[j][i]
			matrix[j][i] = tmp
		}
	}


	for i := 0; i < size; i++ {
		for j := 0; j < size/2; j++ {
			tmp := matrix[i][j]
			matrix[i][j] = matrix[i][size-1-j]
			matrix[i][size-1-j] = tmp
		}
	}
}

func main() {
  var matrix [][]int

  matrix = [][]int{
    []int{1, 2, 3, 4},
    []int{5, 6, 7, 8},
    []int{9, 10, 11, 12},
    []int{13, 14, 15, 16},
  }
  fmt.Println("MATRIX BEFORE: ", matrix)
  rotate(matrix)
  fmt.Println("MATRIX IS NOW: ", matrix)
}
